### PR TITLE
v2

### PR DESCRIPTION
**Pull request by flintinatux**
dilbert

I thought about titling this PR **"Hurray for ADT's!"**, because  that's basically what it's all about.  I heard that `one of our backend projects` was using `paperplane` with ADT's, and in the process was having to wrap all the bits in converter functions to make them work, and... I felt bad.  I mean, that's what `paperplane` is supposed to be about: making backend functional programming easy.  

So, this PR is me fixing that, and making other changes I've been wanting to make, too.

## What's new/changing:
- Support for request handlers that return ADT's (not just Promises) via Redux middleware.
- `mount` is becoming a unary function that accepts a single options object.
- All options have sane defaults (like using `paperplane.logger` as the default `logger`).
- Much simpler error handling via the new `cry` option (and new [`paperplane-airbrake` to match](https://github.com/articulate/paperplane-airbrake/pull/5)).
- Lots of internal cleanup to use other tools that we've extracted in the past year.

For details on the breaking changes, please read the [Migration guide](https://github.com/articulate/paperplane/blob/v2/docs/migration-guide.md).

## How to test:
- [ ] Checkout this branch and https://github.com/articulate/paperplane-airbrake/pull/5 locally.
- [ ] Pick your favorite existing project that already uses `paperplane`.
- [ ] Volume-in the new versions by adding the following to your `docker-compose.override.yml`:
```yml
volumes:
  - '../paperplane:/service/node_modules/paperplane'
  - '../paperplane-airbrake:/service/node_modules/paperplane-airbrake'
```
- [ ] Follow all of the instructions in the [Migration guide](https://github.com/articulate/paperplane/blob/v2/docs/migration-guide.md) to convert your app to v2.
- [ ] Check and make sure everything still works.
- [ ] Then make sure your `APP_ENV=dev`, but set `NODE_ENV=production`.
- [ ] Throw some errors somewhere, and make sure they appear in Airbrake with the request data.
- [ ] Read the updated docs and decide if they make sense.
- [ ] Invite your friends and coworkers to also do the same.

**Special ❤️ note to @evilsoft:**  In case you couldn't tell, this PR is me saying that I want to see more of what you can do in backend apps with `crocks`.  Also, you are specially positioned with <redacted> to really test the new ADT bits in this.  Please do that, because I don't want to ship this until I know it makes your life easier.  And take a look at my function signatures in the docs.  Things got weird when all the legacy bits were Promised-based, but then the signature for the `app` option was `Request -> m Response`.  Let me know how I can make that clearer if need be.

### CODE REVIEW

**Comment by mgreystone** [comment]
I've migrated a few services & all is well. :+1: 

...Why redux middleware? Seems like completely different use cases & implies that things like `thunk` should work.

File: lib/mount.js [hunk 1]
```diff
@@ -1,57 +1,42 @@
...
       .then(parseUrl)
       .then(parseCookies)
       .then(cleanRequest)
-      .then(app)
-      .catch(rethrow(errLogger))
-      .catch(error)
+      .then(handle)
+      .catch(when(
+        is(Error),
+        compose(error, tap(sob(req)))
```
[the remark this answers is not in the document]
Comment (flintinatux) [reply lib/mount.js]: The reason for the `when(is(Error), ...)` is that sometimes in `paperplane` we do intentionally `reject` with an actual Response, not an Error.  The most common example of this is to `Promise.reject` early with a `redirect` response, skipping the rest of the app flow and responding early with a redirect.  This technique is even [mentioned in the docs](https://github.com/articulate/paperplane/blob/master/docs/API.md#redirect), along with an example.

I'd be interested to know what third-party lib you had trouble with, though.

File: lib/mount.js [hunk 2]
```diff
@@ -1,57 +1,42 @@
...
-    return res.end()
-
-  if (body instanceof Stream)
-    return body.pipe(res)
+      .then(() => logger({ req, res }))
 
-  if (!res.getHeader('content-type'))
-    res.setHeader('content-type', 'application/octet-stream')
+  const sob = req => err =>
+    cry(Object.assign(err, { req }))
```
[the remark this answers is not in the document]
Comment (flintinatux) [reply lib/mount.js]: That's correct.  If we `R.merge` onto it, it would lose the stack and everything that makes it an instance of Error.

**Comment by flintinatux** [comment]
@mgreystone, fair question about Redux middleware.  It's sort of a ??? at first glance.

When I started to tackle using `paperplane` as the application "edge" for handling ADT's, a few nice-to-have design requirements began to form:

#### 1. It needs to handle deeply-nested ADT's in a recursive fashion.

For example, much of the time you might just be returning a simple `Async(Error, Response)`.  But other times you might return an `Async(Error, IO(Response))`, or an `IO(Async(Error, Response))`, or any arbitrarily deep-nested combination of ADT's.  So it will need to recursively unwrap all those containers until it gets to the sweet, caramel-y `Response` at the center.

#### 2. It needs to be easily configurable.

Not all ADT implementations work the same.  For example, most `Async`/`Task`/`Future` implementations agree on a `fork` method to execute the work, but for some reason `IO` implementations tend to disagree: some use `run` (like [`crocks`](https://evilsoft.github.io/crocks/docs/crocks/)), others use `runIO`, and then Dr. Boolean uses `unsafePerformIO ` in his [magnum opus on functional programming](https://mostly-adequate.gitbooks.io/mostly-adequate-guide/ch08.html#old-mcdonald-had-effects).

#### 3. It needs to only be enabled if needed.

Apps that never use ADT's shouldn't pay the performance penalty of inspecting responses for potential ADT's.  It should be an opt-in experience, otherwise defaulting to standard Promise chains.

So if you stop and consider those requirements, it turns out that we already have a solution that satisfies all three: **Redux middleware**.  We use them every day to recursively unwrap ADT's, the ones that need to be configurable already are (such as `redux-io`), and they act as plugins, such that if you don't need them, you just don't use them.

And yes, that does mean that `redux-thunk` will actually work.  (When you call the `dispatch` function, it will continue the response flow, and the `getState` function returns a safe `undefined` value.)  However, it's unlikely that I could find a use-case for thunks on the backend.

File: lib/mount.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mgreystone) [reply lib/mount.js]: Oh, it's a _feature_. :confused: I never would have guessed. "Reject" === "Error" in my mind.

The lib where i encountered this was the [chargebee sdk](https://apidocs.chargebee.com/docs/api?lang=node). It uses node-style callbacks which i then promise-ified. But it's first `err` argument is not an `Error` type but a plain object. So when chargebee started responding with errors, paperplane responded with success with a malformed response object, & everything went :boom:

[the remark this answers is not in the document]
Comment (flintinatux) [reply lib/mount.js]: Yuck.  So did you ended doing a `bimap` (or some `Promise` equivalent)?

[the remark this answers is not in the document]
Comment (mgreystone) [reply lib/mount.js]: I ended up reinventing the wheel a bit.

```js
new Promise((resolve, reject) => {
  chargebeeFn((err, result) => {
    if (err != null) {
      reject(Object.assign(new Error(), err))
    }
    resolve(result)
  }
}
```

[the remark this answers is not in the document]
Comment (flintinatux) [reply lib/mount.js]: Bummer.  `funky.reject` will lift a value into an `Error` if it isn't one already, but I'm not sure it would work for objects:
https://github.com/articulate/funky/blob/master/src/reject.js#L7

File: docs/API.md [hunk 3]
```diff
@@ -177,31 +182,66 @@ const app = methods({
...
 ```
 
-Wraps a top-level handler function to prepare for mounting as a new `http` server.  Lifts the handler into a `Promise` chain, so the handler can respond with either a [`Response`](https://github.com/articulate/paperplane/blob/master/docs/getting-started.md#response-object), or a `Promise` that resolves with one.  Also accepts an options object with `errLogger` and `logger` properties, both of which can be set to [`logger`](#logger).
+Wraps a top-level handler function to prepare for mounting as a new `http` server.  Lifts the handler into a `Promise` chain, so the handler can respond with either a synchronous [`Response`](https://github.com/articulate/paperplane/blob/master/docs/getting-started.md#response-object), or a `Promise` or other ADT that resolves with one.  Accepts the following options:
+
+
+| Option | Type | Default | Description |
+| ------ | ---- | ------- | ----------- |
+| `app` | `Request -> m Response` | [`R.identity`](http://devdocs.io/ramda/index#identity) | top-level request handler function |
+| `cry` | `Error -> a` | [`paperplane.logger`](#logger) | error logger |
```
Comment (dbertram) [inline docs/API.md]: Not to be a spoil-sport, but why the change from `errLogger` to `cry`?

`errLogger` is a lot clearer to me in terms of readability/understanding the intended usage without reading the docs.

Comment (flintinatux) [reply docs/API.md]: `cry` is a long-standing idiom on the Rise team for "error reporting", and is already a [function name exported by `glue`](https://github.com/articulate/glue/blob/master/API.md#cry).  I've opted to piggy-back on that custom to borrow the existing connotation.  (And it's likely that `glue#cry` will be dropped in favor of `paperplane-airbrake`, although `glue` will still setup our common Airbrake filters.)

Furthermore, the functionality is changing, since `cry` is intended to be the correct way to report errors to an aggregation service, not just log errors like `errLogger` did.  I thought the functionality change would be less confusing if the name also changed, to prevent a misunderstanding in the purpose.

It's also shorter, saving six whole keystrokes, not to mention that long pinky-stretch to the shift-key.  Your fingers will thank me later.  :trollface:

File: README.md [hunk 4]
```diff
@@ -5,21 +5,23 @@
...
 
 ## Introduction
 
 The main goal of `paperplane` is to make building a `node.js` server easy, without all of the configuration or imperative boilerplate required for other server frameworks.  If you prefer to build apps with function composition or even a point-free style, then `paperplane` is for you.
 
 With `paperplane` you get all of these out-of-the-box:
 
-- pure, functional, Promise-based [route handlers](https://github.com/articulate/paperplane/blob/master/docs/getting-started.md#basic-concepts)
+- pure, functional, Promise-based [request handlers](https://github.com/articulate/paperplane/blob/master/docs/getting-started.md#basic-concepts)
+- customizable support for [request handlers that return ADT's](https://github.com/articulate/paperplane/blob/master/docs/API.md#mount) - **new in v2!**
```
[the remark this answers is not in the document]
Comment (flintinatux) [reply README.md]: Not a bad suggestion.  I think I'll expand the acronym here, and then provide some links to further reading at the bottom of the readme.

File: docs/migration-guide.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+```js
+const cry = compose(console.error, pick(['message', 'name', 'req', 'stack']))
+```
+
+Read more about the new `cry` option [here](./API.md#mount).
+
+### New feature
+
+The `mount` function has gained a new `middleware` option to support request handlers that return ADT's.  To take advantage of this, register a list of appropriate Redux middleware.
```
[the remark this answers is not in the document]
Comment (flintinatux) [reply docs/migration-guide.md]: 👍

**Comment by flintinatux** [comment]
@dbertram, per your suggestion I've expanded some acronyms and added some links for further study on ADT's.  Please let me know if that helps makes things clearer.

**Comment by flintinatux** [comment]
Also, if anyone else wants to put 👀 on this, now's the time.  I think we've let it simmer long enough, and barring any last-minute comments, I plan to ship this shortly.

**Comment by cdwills** [comment]
This looks great @flintinatux .  Would love to eventually get an example in there using ADTs from `crocks` too cc: @evilsoft

**Comment by flintinatux** [comment]
squirrel party
